### PR TITLE
Custom user object

### PR DESCRIPTION
**Pull request by geerteltink**
Resolves zendframework/zend-expressive-authentication#24
Related zendframework/zend-expressive-authentication-session#15

This PR allows a custom user object to be used to store and restore the authenticated user from a session. In stead of adding multiple new methods to the user interface to satisfy every use case, only 2 new methods are added to persist a user in the session.

```php
use Zend\Expressive\Authentication\UserInterface;

class SessionUser implements UserInterface
{
    private $identity;

    private $claims;

    private $roles;

    /**
     * Restore from (session) storage
     *
     * The values here can be customized for your needs, but need to match
     * the keys used in `toState()`.
     */
    public static function fromState(array $data) : UserInterface
    {
        // Validate data here

        $user = new self();

        $user->identity = $data['identity'];
        $user->claims   = (array) $data['claims'];
        $user->roles    = (array) $data['roles'];

        return $user;
    }

    /**
     * Prepare for (session) storage
     *
     * The values here can be customized for your needs, but need to match
     * the keys used in `fromState()`.
     */
    public function toState() : array
    {
        return [
            'identity' => $this->identity,
            'claims'   => $this->claims,
            'roles'    => $this->roles,
        ];
    }

    public function getClaim(string $name, $default = null)
    {
        return $this->claims[$name] ?? $default;
    }

    public function getName() : ?string
    {
        return $this->claims['name'] ?? null;
    }
    
    public function getIdentity() : string
    {
        return $this->identity;
    }

    public function getUserRoles() : array
    {
        return $this->roles;
    }
}
```

- [x] Are you creating a new feature?
  - [x] Why is the new feature needed? What purpose does it serve?
  - [x] How will users use the new feature?
  - [x] Base your feature on the `develop` branch, and submit against that branch.
  - [x] Add only one feature per pull request; split multiple features over multiple pull requests
  - [x] Add tests for the new feature.
  - [ ] Add documentation for the new feature.
  - [ ] Add a `CHANGELOG.md` entry for the new feature.

### CODE REVIEW

File: src/UserRepository/SessionUser.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-authentication for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (https://www.zend.com)
```
Comment (webimpress) [inline src/UserRepository/SessionUser.php]: 2018

File: src/UserRepository/SessionUser.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+declare(strict_types=1);
+
+namespace Zend\Expressive\Authentication\UserRepository;
+
+use Zend\Expressive\Authentication\UserInterface;
+
+class SessionUser implements UserInterface
+{
+    private $identity;
```
Comment (michalbundyra) [inline src/UserRepository/SessionUser.php]: Please add PHP doc with type on the properties (also one below)

File: src/UserRepository/SessionUser.php (hunk 1 above)
Comment (michalbundyra) [inline src/UserRepository/SessionUser.php]: 2018

**Review by ezimuel — COMMENTED** [review]
If you want to add more user's data to `UserInterface` why you don't just create a User class that implements `UserInterface` adding properties and methods as needed? 
Why are you proposing to implement `fromState()` and `toState()`, where you are using it in zend-expressive-authentication?

**Comment by geerteltink** [comment]
> If you want to add more user's data to UserInterface why you don't just create a User class that implements UserInterface adding properties and methods as needed?

I have a class that implements the UserInterface, but that data is not stored in the sessions. Let me clarify this a bit. If I have an OpenID user, I would loose all info except for the identity. All claims are lost etc. But that goes for every custom User class. I would like to see data stored in the session that I choose so I don't need to reload the user every time. 

I have tried this in several other ways, but without hacking into the source code I haven't managed to do that. Unless I'm missing something I can only store the roles and the identity at the moment.

> Why are you proposing to implement fromState() and toState(), where you are using it in zend-expressive-authentication?

It is used in zendframework/zend-expressive-authentication-session#15

**Comment by ezimuel** [comment]
@xtreamwayz zend-expressive-authentication can store any `UserInterface` object in a PSR-7 attribute (see [here](https://github.com/zendframework/zend-expressive-authentication/blob/master/src/AuthenticationMiddleware.php#L36)). 

The point here is: who is generating the `UserInterface` object?
The answer is a `UserRepositoryInterface` object. The goal of this object is to retrieve user's information returning a `UserInterface` instance (by default using [UserTrait](https://github.com/zendframework/zend-expressive-authentication/blob/master/src/UserRepository/UserTrait.php)). This instance can contain any methods/properties that you want but should at least have  the funtions `getIdentity()` and `getUserRoles()`.
I designed the authentication around adapters (e.g. [zend-expressive-authentication-basic](https://github.com/zendframework/zend-expressive-authentication-basic)) and user's repository (that retrieve user info). You need to use a custom user's repository to generate a `UserInterface` object that contains more functions/propoerties.

That said, if you are using [zend-expressive-authentication-oauth2](https://github.com/zendframework/zend-expressive-authentication-oauth2), the usage of `UserRepository` is missing here. I still need to figure out how to apply in it. Right now, it's always returning a User class based on the previous [UserTrait](https://github.com/zendframework/zend-expressive-authentication/blob/master/src/UserRepository/UserTrait.php) (see [here](https://github.com/zendframework/zend-expressive-authentication-oauth2/blob/master/src/OAuth2Adapter.php#L52)).

**Comment by geerteltink** [comment]
> That said, if you are using zend-expressive-authentication-oauth2, the usage of UserRepository is missing here. I still need to figure out how to apply in it. Right now, it's always returning a User class based on the previous UserTrait (see here).

That is exactly what I'm trying to solve with these two PR's. Instead of using the UserTrait, this PR gives us the ability to use a custom User object and allows to persist whatever info that might be needed. The code in this PR is taken from a POC I wrote for OpenID Connect client for single sign on. 

I'm not saying that this is the right solution to solve it, but it definitely gives me the freedom to store an OpenID Connect user with all needed info or I can store the required info in a JWT session token so I don't need to query the database or an API endpoint on every request to grab the missing user info.

If you don't like the solution or have better ideas then please close both PR's and we'll search for a better solution.

**Comment by ezimuel** [comment]
@xtreamwayz I sent #28 as replacement for this PR. I think the solution provided is more general.

**Comment by geerteltink** [comment]
Superseded by zendframework/zend-expressive-authentication#28

**Pull request closed without merging**